File: internal/identity/directory.go
package identity

import "context"

type Directory interface {
	UserExists(ctx context.Context, userID string) (bool, error)
	GroupMembers(ctx context.Context, groupID string) ([]string, error)
	UserGroups(ctx context.Context, userID string) ([]string, error)
}

func IsMember(ctx context.Context, d Directory, userID, groupID string) (bool, error) {
	members, err := d.GroupMembers(ctx, groupID)
	if err != nil {
		return false, err
	}
	for _, m := range members {
		if m == userID {
			return true, nil
		}
	}
	return false, nil
}

type StaticDirectory struct {
	Users  map[string]struct{}
	Groups map[string][]string
}

func NewStaticDirectory(users []string, groups map[string][]string) *StaticDirectory {
	u := make(map[string]struct{}, len(users))
	for _, id := range users {
		u[id] = struct{}{}
	}
	g := make(map[string][]string, len(groups))
	for k, v := range groups {
		cp := make([]string, len(v))
		copy(cp, v)
		g[k] = cp
	}
	return &StaticDirectory{Users: u, Groups: g}
}

func (d *StaticDirectory) UserExists(_ context.Context, userID string) (bool, error) {
	_, ok := d.Users[userID]
	return ok, nil
}

func (d *StaticDirectory) GroupMembers(_ context.Context, groupID string) ([]string, error) {
	members, ok := d.Groups[groupID]
	if !ok {
		return nil, nil
	}
	out := make([]string, len(members))
	copy(out, members)
	return out, nil
}

func (d *StaticDirectory) UserGroups(_ context.Context, userID string) ([]string, error) {
	var out []string
	for gid, members := range d.Groups {
		for _, m := range members {
			if m == userID {
				out = append(out, gid)
				break
			}
		}
	}
	return out, nil
}

File: src/WorkflowEngine/Identity/Directory.cs
namespace WorkflowEngine;

public interface IDirectory
{
    Task<bool> UserExists(string userId, CancellationToken cancellationToken = default);
    Task<IReadOnlyList<string>> GroupMembers(string groupId, CancellationToken cancellationToken = default);
    Task<IReadOnlyList<string>> UserGroups(string userId, CancellationToken cancellationToken = default);
}

public static class DirectoryExtensions
{
    public static async Task<bool> IsMember(this IDirectory directory, string userId, string groupId, CancellationToken cancellationToken = default)
    {
        var members = await directory.GroupMembers(groupId, cancellationToken);
        return members.Contains(userId);
    }
}

public sealed class StaticDirectory : IDirectory
{
    private readonly HashSet<string> _users;
    private readonly Dictionary<string, string[]> _groups;

    public StaticDirectory(IEnumerable<string> users, IReadOnlyDictionary<string, IReadOnlyList<string>> groups)
    {
        _users = new HashSet<string>(users);
        _groups = groups.ToDictionary(kv => kv.Key, kv => kv.Value.ToArray(), StringComparer.Ordinal);
    }

    public Task<bool> UserExists(string userId, CancellationToken cancellationToken = default)
        => Task.FromResult(_users.Contains(userId));

    public Task<IReadOnlyList<string>> GroupMembers(string groupId, CancellationToken cancellationToken = default)
    {
        if (!_groups.TryGetValue(groupId, out var members))
            return Task.FromResult<IReadOnlyList<string>>([]);
        return Task.FromResult<IReadOnlyList<string>>(members.ToArray());
    }

    public Task<IReadOnlyList<string>> UserGroups(string userId, CancellationToken cancellationToken = default)
    {
        var outList = new List<string>();
        foreach (var (gid, members) in _groups)
        {
            if (members.Contains(userId))
                outList.Add(gid);
        }
        return Task.FromResult<IReadOnlyList<string>>(outList);
    }
}
